orttraining/orttraining/training_api/optimizer.cc

```diff
@@ -40,53 +49,93 @@ Status Optimizer::ConstructInputs() {
   if (optimizer_type_ == OptimizerType::AdamW) {
     auto& param_named_optimizer_states = optimizer_state_.param_named_optimizer_states;
 
-    std::string param_name;
-    std::vector<std::string> param_names, grad_names, moment1_names, moment2_names, user_inputs;
-    for (size_t i = 2; i < input_names_.size(); i++) {
-      std::string& name = input_names_[i];
-      auto it = named_parameters_.find(name);
-      if (it != named_parameters_.end()) {  // is param
-        param_names.push_back(name);
-        inputs_.push_back(it->second->Data());
-      } else if (utils::GetParamNameFromGradient(name, param_name)) {
-        grad_names.emplace_back(name);
-        // assert param_name is valid.
-        auto it = named_parameters_.find(param_name);
-        ORT_ENFORCE(it != named_parameters_.end(), "Unknown param: ", param_name, " for field: ", name);
-        inputs_.push_back(it->second->Gradient());
-      } else if (utils::GetParamNameFromSuffix(name, MOMENT_SUFFIXES[0], param_name)) {
-        moment1_names.push_back(name);
-        auto it = named_parameters_.find(param_name);
-        ORT_ENFORCE(it != named_parameters_.end(), "Unknown param: ", param_name, " for field: ", name);
-        inputs_.push_back(param_named_optimizer_states.at(param_name).momentum_named_states.at(MOMENT_STATE_NAMES[0]));
-      } else if (utils::GetParamNameFromSuffix(name, MOMENT_SUFFIXES[1], param_name)) {
-        moment2_names.push_back(name);
-        auto it = named_parameters_.find(param_name);
-        ORT_ENFORCE(it != named_parameters_.end(), "Unknown param: ", param_name, " for field: ", name);
-        inputs_.push_back(param_named_optimizer_states.at(param_name).momentum_named_states.at(MOMENT_STATE_NAMES[1]));
+    std::vector<Tensor> params, first_order_moments, second_order_moments;
+    // TODO: Change to tensor seq implementation once clip grad norm op
+    // that accepts tensor seq as input for gradients is complete.
+    std::vector<OrtValue> grads;
+
+    // Collect all the params, grads, and moments based on grad names
+    // Input names 0-4 are reserved for lr, step, params, first order moments, second order moments
+    // input names 5 onwards are all the gradient names.
+    // Collect all the inputs based on the gradient names order.
+    for (size_t i = 5; i < input_names_.size(); i++) {
+      std::string param_name;
+      if (utils::GetParamNameFromGradient(input_names_[i], param_name)) {
```

Comment:
Eventually when we have param/grad/states all in seq<tensor> as graph inputs. actually, there will not be parameter names given in the optimizer graph, right? So, I am thinking, during Optimizer construction, we should just iterate all trainable parameters from Module, and feed them in param/grads/states in the same order of named_parameters of Module.

Reply:
Yes, we just use the order from the `named_parameters_` eventually. For now, we need to do this because the order of grads is important for inputs.